lib/projfs.c

```diff
@@ -77,6 +79,62 @@ static inline int get_fuse_context_lowerdir_fd(void)
 	return get_fuse_context_projfs()->lowerdir_fd;
 }
 
+// ceil(log10(INT_MAX)) = ceil(log10(2) * sizeof(int) * CHAR_BIT)
+//			<     (   1/3   * sizeof(int) * CHAR_BIT) + 1
+#define INT_FMT_LEN ((sizeof(int) * CHAR_BIT) / 3 + 1)
+
+#define PROC_STATUS_PATH_FMT "/proc/%d/status"
+#define MAX_PROC_STATUS_PATH_LEN \
+	(sizeof(PROC_STATUS_PATH_FMT) + INT_FMT_LEN - 3)
+
+#define PROC_STATUS_BUF_SIZE 32
+
+#define PROC_STATUS_TGID_KEY "Tgid:"
+#define PROC_STATUS_TGID_KEY_LEN (sizeof(PROC_STATUS_TGID_KEY) - 1)
+
+// NOTE: only functional within a FUSE file operation!
+static pid_t get_fuse_context_tgid(void)
+{
+	pid_t pid = fuse_get_context()->pid;
+	char path[MAX_PROC_STATUS_PATH_LEN + 1];
+	char buf[PROC_STATUS_BUF_SIZE];
+	FILE *file;
+	int found = 0;
+
+	// do not report IO or parsing errors
+	sprintf(path, PROC_STATUS_PATH_FMT, pid);
+	file = fopen(path, "r");
+	if (file == NULL)
+		return pid;		// best effort
+	while (fgets(buf, sizeof(buf), file) != NULL) {
+		if (strncmp(buf, PROC_STATUS_TGID_KEY,
+			    PROC_STATUS_TGID_KEY_LEN) == 0)
+		{
+			found = 1;
+			break;
+		}
+		while (strchr(buf, '\n') == NULL &&
+		       fgets(buf, sizeof(buf), file) != NULL);
```

Comment:
Very nice!

Reply:
Thanks -- sometimes the gods grant small mercies.